library/psa_crypto_se.c

```diff
@@ -75,6 +90,123 @@ const psa_drv_se_t *psa_get_se_driver( psa_key_lifetime_t lifetime )
 
 
 
+/****************************************************************/
+/* Slot management */
+/****************************************************************/
+
+/** The type of bit vector elements. A bit vector is represented as an
+ * array of bit vector elements. */
+typedef unsigned bv_element_t;
+/* Number of bits per bit vector elements. Assumes no padding bits,
+ * which Mbed TLS checks in selftest.c. */
+#define BV_BITS_PER_ELEMENT ( sizeof( bv_element_t ) * CHAR_BIT )
+
+struct psa_drv_se_slot_usage_s
+{
+    size_t size; /** Number of elements of \c vector */
+    bv_element_t vector[1]; /* Should be vector[] but we don't allow C99 */
+};
+
+static psa_key_slot_number_t su_size(
+    const psa_drv_se_slot_usage_t *slot_usage )
+{
+    return( slot_usage->size * BV_BITS_PER_ELEMENT );
+}
+
+static int su_get( const psa_drv_se_slot_usage_t *slot_usage,
+                   psa_key_slot_number_t n )
+{
+    bv_element_t elt = slot_usage->vector[n / BV_BITS_PER_ELEMENT];
+    return( elt >> (n % BV_BITS_PER_ELEMENT) & 1 );
+}
+
+static void su_set( psa_drv_se_slot_usage_t *slot_usage,
+                    psa_key_slot_number_t n,
+                    int value )
+{
+    bv_element_t *elt = &slot_usage->vector[n / BV_BITS_PER_ELEMENT];
+    bv_element_t mask = 1 << (n % BV_BITS_PER_ELEMENT);
+    if( value )
+        *elt = *elt | mask;
+    else
+        *elt = *elt & ~mask;
+}
+
+/* Callback function for drivers */
+psa_status_t psa_drv_cb_find_free_slot(
+    const psa_drv_se_slot_usage_t *slot_usage,
+    psa_key_slot_number_t from,
+    psa_key_slot_number_t before,
+    psa_key_slot_number_t *found )
+{
+    psa_key_slot_number_t n;
+    if( from >= su_size( slot_usage ) )
+        return( PSA_ERROR_INSUFFICIENT_STORAGE );
+    if( before > su_size( slot_usage ) )
+        before = su_size( slot_usage );
+    for( n = from; n < before; n++ )
+    {
+        if( ! su_get( slot_usage, n ) )
+        {
+            *found = n;
+            return( PSA_SUCCESS );
+        }
+    }
+    return( PSA_ERROR_INSUFFICIENT_STORAGE );
+}
+
+static psa_status_t load_slot_usage( psa_se_drv_table_entry_t *drv )
+{
+    size_t n_elements = ( drv->methods->key_management->slot_count +
+                          BV_BITS_PER_ELEMENT - 1 ) / BV_BITS_PER_ELEMENT;
+    size_t size = ( sizeof( *drv->slot_usage )
+                    - sizeof( drv->slot_usage->vector )
+                    + sizeof( drv->slot_usage->vector[0] ) * n_elements );
+    drv->slot_usage = mbedtls_calloc( 1, size );
+    if( drv->slot_usage == NULL )
+        return( PSA_ERROR_INSUFFICIENT_MEMORY );
+    drv->slot_usage->size = drv->methods->key_management->slot_count;
+
+    /* TOnogrepDO: load */
```

Comment:
Leftover comment

Reply:
No, it's still to do.

Reply:
Should this PR be blocked on loading not being done yet?

Reply:
It shouldn't be merged as is because it includes a feature that's on by default (SE support) but leads to undefined behavior if used according to the documentation. I don't have precise criteria in mind for when this PR will be mergeable: ideally it would do everything but I also want it to be usable ASAP.

Reply:
I'd recommend not making this PR do everything as a precondition to merging.

Let's try to pick a safe subset to enable by default, and implement that bit. Or, make this off by default until we are comfortable enough with the feature to turn it on by default. The latter approach is quite common and widely known in the industry as a "feature toggle".

Reply:
I'm not planning to remove TOnogrepDO's from this PR in general. This particular one has actually been done.